phoenix-core/src/main/java/org/apache/phoenix/coprocessor/PhoenixTTLRegionObserver.java

```diff
@@ -0,0 +1,368 @@
+                return true;
+            }
+
+            Iterator<Cell> cellIterator = cellList.iterator();
+            Cell firstCell = cellIterator.next();
+            byte[] rowKey = new byte[firstCell.getRowLength()];
+            System.arraycopy(firstCell.getRowArray(), firstCell.getRowOffset(), rowKey, 0,
+                    firstCell.getRowLength());
+
+            boolean isRowExpired = !checkRowNotExpired(cellList);
+            if (isRowExpired) {
+                long ttl = ScanUtil.getPhoenixTTL(this.scan);
+                long ts = getMaxTimestamp(cellList);
+                LOG.debug(String.format(
+                        "** PHOENIX-TTL: Deleting region = %s, row = %s, delete-ts = %d, max-ts = %d ** ",
+                        region.getRegionInfo().getTable().getNameAsString(), Bytes.toString(rowKey),
+                        now - ttl, ts));
+                Delete del = new Delete(rowKey, now - ttl);
+                Mutation[] mutations = new Mutation[] { del };
+                region.batchMutate(mutations, HConstants.NO_NONCE, HConstants.NO_NONCE);
+                return true;
+            }
+            return false;
+        }
+
+        private boolean isEmptyColumn(Cell cell) {
+            return Bytes.compareTo(cell.getFamilyArray(), cell.getFamilyOffset(),
+                    cell.getFamilyLength(), emptyCF, 0, emptyCF.length) == 0 &&
+                    Bytes.compareTo(cell.getQualifierArray(), cell.getQualifierOffset(),
+                            cell.getQualifierLength(), emptyCQ, 0, emptyCQ.length) == 0;
+        }
+
+        // TODO : Remove it after we verify all SQLs include the empty column.
+        // Before we added ScanUtil.addEmptyColumnToScan
+        // some queries like select count(*) did not include the empty column in scan,
+        // thus this method was the fallback in those cases.
+        private boolean checkEmptyColumnNotExpired(byte[] rowKey) throws IOException {
+            LOG.warn("Scan " + scan + " did not return the empty column for " + region
+                    .getRegionInfo().getTable().getNameAsString());
+            Get get = new Get(rowKey);
+            get.setTimeRange(minTimestamp, maxTimestamp);
+            get.addColumn(emptyCF, emptyCQ);
+            Result result = region.get(get);
+            if (result.isEmpty()) {
+                LOG.warn("The empty column does not exist in a row in " + region.getRegionInfo()
+                        .getTable().getNameAsString());
+                return false;
+            }
+            return !isTTLExpired(result.getColumnLatestCell(emptyCF, emptyCQ));
+        }
+
+        /**
+         * @param cellList is an input and output parameter and will either include a valid row
+         *                 or be an empty list
+         * @return true if row not expired, otherwise false
+         * @throws IOException
+         */
+        private boolean checkRowNotExpired(List<Cell> cellList) throws IOException {
```

Comment:
It would be great if we could have unit tests for all these new methods inside say a new file `PhoenixTTLRegionObserverTest.java`

Reply:
@jpisaac can we please add these tests?